lib/client.ts

```diff
@@ -604,22 +613,54 @@ export class ServiceClient {
 
     const retryErrors: ServiceClientError[] = [];
     return new Promise<ServiceClientResponse>((resolve, reject) => {
+      const timerInitial = performance.now();
       const breaker = this.getCircuitBreaker(params);
-      const retryOperation = operation(opts, (currentAttempt: number) => {
+      const requestsAbortCallbacks: (() => void)[] = [];
+      const retryOperation = operation(opts, scheduledRetry => {
         breaker.run(
           (success: () => void, failure: () => void) => {
+            if (this.options.dropAllRequestsAfter) {
+              const timerLeft =
+                this.options.dropAllRequestsAfter -
+                (performance.now() - timerInitial);
+              if (params.dropRequestAfter) {
+                params.dropRequestAfter = Math.min(
+                  params.dropRequestAfter,
+                  timerLeft
+                );
+              } else {
+                params.dropRequestAfter = timerLeft;
+              }
+            }
+
+            if (retryAfter) {
+              params.registerAbortCallback = cb =>
+                requestsAbortCallbacks.push(cb);
+            }
+
             return requestWithFilters(
               this,
               params,
               this.options.filters || [],
               this.options.autoParseJson
             )
               .then((result: ServiceClientResponse) => {
+                if (retryOperation.isResolved()) {
+                  return;
+                }
+                retryOperation.resolved();
                 success();
                 result.retryErrors = retryErrors;
                 resolve(result);
+                if (retryAfter) {
+                  requestsAbortCallbacks.forEach(cb => cb());
+                  requestsAbortCallbacks.length = 0;
```

Comment:
So, this is a cleanup for the happy case. Is this array of callbacks cleaned up in the sad case (all requests fail)? I'm not sure if it causes a memory leak only from the code, but I think it's worth checking.

Reply:
Yes good catch!